internal/pkg/clock/chainclock.go

```diff
@@ -25,21 +19,41 @@ type chainClock struct {
 	// GenesisTime is the time of the first block. EpochClock counts
 	// up from there.
 	GenesisTime time.Time
+	// EpochDuration is the time an epoch takes
+	EpochDuration time.Duration
 
 	Clock
 }
 
-// NewChainClock returns a ChainEpochClock.
-func NewChainClock(genesisTime uint64) ChainEpochClock {
-	gt := time.Unix(int64(genesisTime), 0)
-	return &chainClock{GenesisTime: gt, Clock: NewSystemClock()}
+// NewChainClock returns a ChainEpochClock wrapping a default clock.Clock
+func NewChainClock(genesisTime uint64, blockTime time.Duration) ChainEpochClock {
+	return NewChainClockFromClock(genesisTime, blockTime, NewSystemClock())
+}
+
+// NewChainClockFromClock returns a ChainEpochClock wrapping the provided
+// clock.Clock
+func NewChainClockFromClock(genesisTime uint64, blockTime time.Duration, c Clock) ChainEpochClock {
+	gt := time.Unix(int64(genesisTime), int64(genesisTime%1000000000))
```

Comment:
I've been thinking about this.  While sub-second precision was helpful for testing I don't think its necessary using mock clocks.  I became aware working on the next PR that the timestamp only records up to the second.  With that in mind subsecond epochs really don't make sense and a non-second-aligned genesis also doesn't make sense.

Reply:
In retrospect this is a massive headache for integration testing.  We need to be able to run networks with subsecond blocktimes but the timestamps can't differentiate epochs.  

As a result we need `EpochAtTime` to return a range of epochs for validation instead of a single epoch

Reply:
Many integration tests can use fake clocks too. Only forking real processes will require a whole-second clock.